### ISSUE TITLE
Improving cheating

### ISSUE THREAD

**Issue by Drup** [dbuenzli/cmdliner#36]
As you know, in functoria, I'm using `Term.eval_peek_ops` heavily to cheat my way around applicativeness. I know it's bad, but I don't have the choice. Unfortunately, controlling eval_peek_ops has proven very hard in two directions: 
- keeping the documentation decent
- Having good error reporting

The first one can mostly be solved by following this recipe: have a strict inclusion relation between what you peek and what's inside the term at the end. (It's a good idea anyway, for various reasons ...)

However, I haven't managed to solve the other one. Let's consider the following scenario: I have an optional term `t` resolving to the name of a config file, I peek it, if it's present I use it. Then I try to compile/link the config file, if it works, I use it to generate a term and `eval` it. However, if it doesn't, I would like to fail immediately. Unfortunately, if there is on the command line an option that is only in the generated term, cmdliner will say that the option is unknown, not that the compilation failed!

I think I see several way to fix that:
- Implement proper chaining, that propagate results along the way correctly
- Add a `fail` term, that always fail, and accepts everything on the command line.
- Add an exception that deliver a `Term.ret` directly. (meh)

**Comment by dbuenzli** [comment]
Your second proposal of fixing. Can already be implemented by using `Term.ret` so I'm not sure there's anything special to be done here.

**Comment by Drup** [comment]
You mean that `Term.(ret (pure `Error ("bla",b)))` cannot fail because of mismatched arguments ? I'm surprised by that ...

**Comment by Drup** [comment]
(I would really prefer the first option anyway, and it would make usage of peek much clearer)

**Comment by dbuenzli** [comment]
I don't understand what the first option means. Do what you mention and make sure that it's the first term that you evaluate when you construct the larger term, it should work, evaluation stops as soon as an error occurs.

Considering that we are really pushing the system to its limits here I'm not sure it's worth introducing new ways for now.

**Comment by dbuenzli** [comment]
it will however answer the `--help` option even if there's the failing term --- which is right and you'll get the docs of whatever was in the larger term.

**Comment by dbuenzli** [comment]
(and the `--version` if you required it).

**Comment by Drup** [comment]
Except it doesn't do what you are saying: 

```ocaml
open Cmdliner ;;
let t = Term.(ret @@ pure (`Error (false, "error"))) ;;
Term.eval ~argv:[| "bla" ; "bli" ; "blo" |] (t, Term.info "") ;;
```

returns 
```
: too many arguments, don't know what to do with `bli', `blo'                   
Usage:  [OPTION]...                                                             
Try ` --help' for more information.                                             
- : 'a Term.result = `Error `Parse
````

**Comment by Drup** [comment]
(And it's consistent really: parsing before evaluation. If parsing fails, evaluation is not even considered ...)

**Comment by dbuenzli** [comment]
Ok so apparently I misunterpreted what I did one year ago, and this is actually documented [here](https://github.com/samoht/assemblage/blob/master/lib-driver/assemblage_driver.mli#L181-L183) (see the right hand side of the *or* clause). In any case I don't see why you can't output the compilation errors before so that at least [this](https://github.com/mirage/mirage/issues/461) doesn't happen and I'm not sure what the issue is [here](https://github.com/mirage/mirage/issues/462).

Here's what I have currently in another system and while it may not be the clearest you can at least see that there is some error cascading issue. 
```
> ./b0 --bla
File "_build/b0/build.ml", line 3, characters 1-2:
Error: Syntax error
b0: unknown option `--bla'.
Usage: b0 [OPTION]... 
Try `b0 --help' for more information.
```

I will improve this in the future, but not now.

**Comment by Drup** [comment]
Ok, so you got exactly the same issue than I do, but since you know cmdliner better, your designed your system around it. Fabulous.

**Comment by dbuenzli** [comment]
Well you seem to be idiotic enough to actually hide to your users that they have syntax errors in their config files. Bravo.

**Comment by Drup** [comment]
Nope, I waited for your answer to this ticket before introducing hacks into my codebase in the hope for a clean solution *shrug*.

**Comment by dbuenzli** [comment]
> I waited for your answer to this ticket before introducing hacks into my codebase in the hope for a clean solution.

You should plan a little bit better. You had all summer to talk me about this and I worked on a new release of cmdliner in the beginning of october. Sorry I'm not going back to this right now I have other things to do.

**Comment by dbuenzli** [comment]
@Drup rereading this I'm not exactly sure what you are asking for. In particular I don't see the problem with: 

> Let's consider the following scenario: I have an optional term t resolving to the name of a config file, I peek it, if it's present I use it. Then I try to compile/link the config file, if it works, I use it to generate a term and eval it. However, if it doesn't, I would like to fail immediately. Unfortunately, if there is on the command line an option that is only in the generated term, cmdliner will say that the option is unknown, not that the compilation failed!

can't you simply not eval or eval something else if the compilation fails ?

**Comment by Drup** [comment]
I think @hannesm and/or @yallop can explain it a bit better than I would, they fought with functoria `-f` too.

But basically, that's what we (used to) do, but it's just so extremely error prone to do it by hand, to the point that the option was dropped in mirage 3.0. I think it would be much much easier to have one function in cmdliner that does the chaining.

**Comment by dbuenzli** [comment]
> I think it would be much much easier to have one function in cmdliner that does the chaining.

I don't understand what this "chaining" means. 

That said I'm not really happy about this API and I may deprecate it at some point in favor of something better now that the internals are a little bit cleaned up and more clear.

**Comment by hannesm** [comment]
Talking about functoria (which does ocamlbuild + dynload to find additional parameters) and `--help`, the problem is that `Term.eval_peek_opts` returns always `` `Help ``  if `--help` is part of `Sys.argv` (amongst other options).  You cannot extract these options from `argv` using `eval_peek_opts`.  See https://github.com/mirage/functoria/pull/97#issuecomment-268563346 for further information about functoria and cmdliner.

By now, `-f` was removed from functoria (I personally don't have any bad feelings about it, I don't think it makes sense to provide this degree of freedom.

**Comment by dbuenzli** [comment]
> Term.eval_peek_opts returns always `Help if --help is part of Sys.argv (amongst other options).

Well the first pair of the component returned by `eval_peek_opts` precisely returns what would happen if you ignore `--help`...

**Comment by dbuenzli** [comment]
Bon since I don't see what cannot be achieved in @Drup's initial scenario with `eval_peek_opts` or what the problem is I'm going to close this (you may want to read the [documentation again](http://erratique.ch/software/cmdliner/doc/Cmdliner.Term.html#VALeval_peek_opts)). The dupe documentation problem that @Drup mentions has been solved in #77. 

Feel free to open a new issue whenever you are able to describe me exactly the problem, why it cannot be solved by `eval_peek_opts` and what you propose instead.
